internal/data/metric.go

```diff
@@ -65,42 +77,145 @@ func (md MetricData) MetricCount() int {
+		// User may have changed internal fields in any ResourceMetrics, flush all of them.
+		md.orig = make([]*otlpmetrics.ResourceMetrics, len(md.pimpl.resourceMetrics))
+		for i := range md.pimpl.resourceMetrics {
+			md.orig[i] = md.pimpl.resourceMetrics[i].getOrig()
+			md.pimpl.resourceMetrics[i].flushInternal()
+		}
+	} else {
+		// User may have changed internal fields in any ResourceMetrics, flush all of them.
+		for i := range md.pimpl.resourceMetrics {
+			md.pimpl.resourceMetrics[i].flushInternal()
+		}
+	}
+	return md.orig
+}
+
+func (md MetricData) initInternallIfNeeded() {
+	if !md.pimpl.initialized {
+		md.pimpl.resourceMetrics = newResourceMetricsSliceFromOrig(md.orig)
+		md.pimpl.initialized = true
+	}
+}
+
 // ResourceMetrics is a collection of metrics from a Resource.
 //
 // Must use NewResourceMetrics functions to create new instances.
 // Important: zero-initialized instance is not valid for use.
 type ResourceMetrics struct {
-	orig *otlpmetric.ResourceMetrics
+	orig *otlpmetrics.ResourceMetrics
 
 	// Override a few fields. These fields are the source of truth. Their counterparts
 	// stored in corresponding fields of "orig" are ignored.
 	pimpl *internalResourceMetrics
 }
 
 type internalResourceMetrics struct {
-	resource *Resource
+	resource Resource
 	metrics  []Metric
+	// True when the slice was replace.
+	sliceChanged bool
+	// True if the pimpl was initialized.
+	initialized bool
 }
 
 // NewResourceMetrics creates a new ResourceMetrics.
 func NewResourceMetrics() ResourceMetrics {
-	return ResourceMetrics{&otlpmetric.ResourceMetrics{}, &internalResourceMetrics{}}
+	return ResourceMetrics{&otlpmetrics.ResourceMetrics{}, &internalResourceMetrics{}}
+}
+
+// NewResourceMetricsSlice creates a slice of ResourceMetrics that are correctly initialized.
+func NewResourceMetricsSlice(len int) []ResourceMetrics {
+	// Slice for underlying orig.
+	origs := make([]otlpmetrics.ResourceMetrics, len)
+	// Slice for underlying pimpl.
+	pimpls := make([]internalResourceMetrics, len)
+	// Slice for wrappers.
+	wrappers := make([]ResourceMetrics, len)
+	for i := range origs {
+		wrappers[i].orig = &origs[i]
+		wrappers[i].pimpl = &pimpls[i]
+	}
+	return wrappers
+}
+
+func newResourceMetricsSliceFromOrig(origs []*otlpmetrics.ResourceMetrics) []ResourceMetrics {
+	// Slice for underlying pimpl.
+	pimpls := make([]internalResourceMetrics, len(origs))
+	// Slice for wrappers.
+	wrappers := make([]ResourceMetrics, len(origs))
+	for i := range origs {
+		wrappers[i].orig = origs[i]
+		wrappers[i].pimpl = &pimpls[i]
+	}
+	return wrappers
 }
 
-func (rm ResourceMetrics) Resource() *Resource {
+func (rm ResourceMetrics) Resource() Resource {
+	rm.initInternallIfNeeded()
 	return rm.pimpl.resource
 }
 
-func (rm ResourceMetrics) SetResource(r *Resource) {
+func (rm ResourceMetrics) SetResource(r Resource) {
+	rm.initInternallIfNeeded()
 	rm.pimpl.resource = r
+	rm.orig.Resource = r.getOrig()
```

Comment:
Any reason we do not use "changed" flag for Resource?

Reply:
Because it has a map, and I don't know when a change happened, so I always have to sync it.

Reply:
Hmm, I may be missing something, but I think attribute map changes should be tracked by `Resource` itself. Here we only need to track if entire Resource was reassigned. For example if we have this:
```go
rm := &ResourceMetrics{...}
rm.Resource().Attributes()["abc"]="def"
```
We do not need to set "changed" flag for `ResourceMetrics`, we only need to set "changed" flag for `Resource` when `Attributes()` is called.

A further improvement could be if we encapsulate attribute map as we discussed so that you have to use `SetAttribute` function explicitly to change an attributes and which will set the "changed" flag. In that case `Attributes()` function will not exist. But even without this improvement the "changed" flag can be tracked by each struct within its area of responsibility. I think ResourceMetrics implementation should not care about or depend on Resource having or not having a map field inside.

Reply:
For resource I don't need for the moment to track any change because if it is created I always reconstruct the otlp representation anyway. So if the resource changed via "SetResource" the object in pimpl will be the new one and I will call `toOrig`